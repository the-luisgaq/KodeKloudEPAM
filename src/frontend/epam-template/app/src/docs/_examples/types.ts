import { TTypeProp } from '@epam/uui-docs';

export type ExampleProps = {
    propDocs?: Record<string, TTypeProp>;
};
