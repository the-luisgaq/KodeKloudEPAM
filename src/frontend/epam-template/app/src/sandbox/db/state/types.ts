import { Person, PersonEmploymentGroup } from '@epam/uui-docs';

export type PersonTableRecord = Person | PersonEmploymentGroup;
