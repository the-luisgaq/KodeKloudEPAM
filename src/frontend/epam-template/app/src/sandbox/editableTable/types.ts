import { DropPosition } from '@epam/uui-core';

export interface Task {
    id: number;
    parentId?: number;
    name: string;
    estimate?: number;
    resources?: number[];
    startDate?: string;
    dueDate?: string;
    status?: string;
    description?: string;
    order?: string;
}

export interface Resource {
    id: number;
    name: string;
    fullName: string;
}

export interface Status {
    id: number;
    name: string;
    color?: string;
}

export type InsertTaskCallback = (position: DropPosition, relativeTask?: Task | null, existingTask?: Task | null) => void;
export type DeleteTaskCallback = (task: Task) => void;

export interface ColumnsProps {
    insertTask: InsertTaskCallback;
    deleteTask: DeleteTaskCallback;
}
