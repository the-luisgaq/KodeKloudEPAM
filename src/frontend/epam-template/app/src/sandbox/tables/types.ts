import { DataQueryFilter } from '@epam/uui-core';
import { Person, PersonEmploymentGroup, Location } from '@epam/uui-docs';
import { DataSourceState, LazyDataSourceApiResponse } from '@epam/uui-core';

export type PersonTableRecord = Person | PersonEmploymentGroup | Location;

export type PersonTableRecordType = PersonTableRecord['__typename'];

export type PersonTableRecordId = [PersonTableRecordType, string | number];

export type PersonTableId = [];

export type PersonTableFilter = DataQueryFilter<Person> & { groupBy?: string };

export interface PersonsTableState extends DataSourceState {
    isFolded?: boolean;
}

export interface PersonsApiResponse extends LazyDataSourceApiResponse<Person | PersonEmploymentGroup> {
    summary: PersonsSummary;
    totalCount: number;
}

export interface PersonsSummary extends Pick<PersonsApiResponse, 'totalCount'> {
    totalSalary: string;
}
