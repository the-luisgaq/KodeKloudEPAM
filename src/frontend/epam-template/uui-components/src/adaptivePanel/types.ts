import { IHasCX, IHasRawProps } from '@epam/uui-core';

export type AdaptiveItemProps<T = unknown> = T & {
    /**
     * Render callback of the item. It renders items inside the panel and measures their width.
     * Pay attention that if you want to set some gaps between items, use AdaptivePanel property itemsGap.
     * */
    render: (item: AdaptiveItemProps<T>, hiddenItems?: AdaptiveItemProps<T>[], displayedItems?: AdaptiveItemProps<T>[]) => any;
    /**
     * Item collapsing priority. Items with lower priority will be hidden first.
     * If several items have the same priority, they will be hidden together, even if there's a place for one of them.
     * */
    priority: number;
    /** If true, this item will be shown when some other items was hidden; for example, you can use it to render dropdowns with hidden items.
     * You can provide more than one collapsedContainer item, but will be shown only those which has minimal priority, but this priority can't be less than the last hidden itemâ€™s priority. */
    collapsedContainer?: boolean;
    /** Unique ID of item */
    id: string;
};

export interface AdaptivePanelProps extends IHasCX, IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    /** Array of items to be rendered in AdaptivePanel */
    items: AdaptiveItemProps[];
    /** Defines size of the gap (gutter) between an element's 'items' in the AdaptivePanel */
    itemsGap?: number | '6' | '12' | '18' | '24' | '36';
}
