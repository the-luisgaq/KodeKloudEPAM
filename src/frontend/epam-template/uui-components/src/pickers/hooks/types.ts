import {
    DataSourceState, IEditable, IHasCaption, IHasRawProps, IModal, PickerBaseOptions, PickerBaseProps,
    PickerFooterProps, DataRowProps, SortingOption,
} from '@epam/uui-core';
import { Dispatch, ReactNode, SetStateAction } from 'react';

export interface UsePickerInputStateProps extends UsePickerStateProps {}

export interface PickerInputState extends PickerState {
    opened: boolean;
    setOpened: Dispatch<SetStateAction<boolean>>;
    isSearchChanged: boolean;
    setIsSearchChanged: Dispatch<SetStateAction<boolean>>;
}

export interface UsePickerStateProps {
    dataSourceState?: Partial<DataSourceState>;
}

export interface PickerState {
    dataSourceState: DataSourceState;
    setDataSourceState: Dispatch<SetStateAction<DataSourceState>>;
    showSelected: boolean;
    setShowSelected: Dispatch<SetStateAction<boolean>>;
}

export type PickerListBaseProps<TItem, TId> = PickerBaseProps<TItem, TId> & {
    /**
     * Number of default items to show initially, when nothing is selected.
     * @default 10
     */
    maxDefaultItems?: number;

    /**
     * Maximum total number of items to show, including selected
     * @default 50
     */
    maxTotalItems?: number;

    /**
     * Ids of items to show first.
     * If not specified, top props.maxDefaultItems will be shown according to the DataSource sorting settings (default is 10)
     */
    defaultIds?: TId[];

    /** If provided, top picks will be automatically adjusted based on last user selection, and stored as user setting under provided key */
    settingsKey?: string;

    sortBy?(item: TItem, sorting: SortingOption): string;

    /** Render callback for list row.
     * For modal row, use renderRow prop.
     *
     * If omitted, PickerListRow component will be rendered.
     */
    renderListRow?: (props: DataRowProps<TItem, TId>) => ReactNode;
};

export type UsePickerListProps<TItem, TId, TProps> = PickerListBaseProps<TItem, TId> & TProps & {};

export interface PickerModalOptions<TItem, TId> extends IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    /** Render callback for custom filters block. This block will be rendered befor the items list.
     If omitted, nothing will be rendered.
     */
    renderFilter?(editableFilter: IEditable<any>): React.ReactNode;
    /** Render callback for modal footer.
     If omitted, default footer will be rendered.
     */
    renderFooter?: (props: PickerFooterProps<TItem, TId> & Partial<IModal<any>>) => React.ReactNode;
    /**
     * If true, prevent modal window closing by click outside modal
     * @default false
     */
    disallowClickOutside?: boolean;
}

export type PickerModalScalarProps<TId, TItem> =
    | ({ selectionMode: 'single'; valueType: 'id'; initialValue: TId } & IModal<TId>)
    | ({ selectionMode: 'single'; valueType: 'entity'; initialValue: TItem } & IModal<TItem>);
export type PickerModalArrayProps<TId, TItem> =
    | ({ selectionMode: 'multi'; valueType: 'id'; initialValue: TId[] } & IModal<TId[]>)
    | ({ selectionMode: 'multi'; valueType: 'entity'; initialValue: TItem[] } & IModal<TItem[]>);

export type UsePickerModalProps<TItem, TId> = PickerBaseOptions<TItem, TId> &
IHasCaption &
(PickerModalScalarProps<TId, TItem> | PickerModalArrayProps<TId, TItem>) &
PickerModalOptions<TItem, TId>;
