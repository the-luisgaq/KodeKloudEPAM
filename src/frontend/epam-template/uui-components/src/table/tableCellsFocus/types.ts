import { DataTableCellProps } from '@epam/uui-core';
import type { DataTableFocusManager } from './DataTableFocusManager';

export type RowsRegistry<TId> = Map<TId | string, Array<CellProps>>;

export type CellProps = Pick<DataTableCellProps, 'index' | 'key' | 'isReadonly' | 'isDisabled'> & CellActions;

export interface CellActions {
    focus: () => void;
}

export interface RowInfo<TId> {
    id: TId,
    index: number;
}

export type { DataTableFocusManager };
