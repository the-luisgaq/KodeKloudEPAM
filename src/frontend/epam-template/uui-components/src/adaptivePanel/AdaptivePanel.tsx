import React, { useEffect, useRef, useState } from 'react';
import cx from 'classnames';
import { useLayoutEffectSafeForSsr } from '@epam/uui-core';
import { measureAdaptiveItems } from './measureItemsUtils';
import type { AdaptivePanelProps } from './types';

import css from './AdaptivePanel.module.scss';

export function AdaptivePanel(props: AdaptivePanelProps) {
    const [itemsWidth, setItemsWidth] = useState<Record<string, number>>();
    const [isChanged, setIsChanged] = useState(false);
    const wrapperRef = useRef<HTMLDivElement>(null);
    const displayedRowRef = useRef<HTMLDivElement>(null);
    const itemsGap = props.itemsGap ? Number(props.itemsGap) : 0;

    const getItemsWidth = () => {
        if (!displayedRowRef.current) {
            return;
        }
        const children = Array.from(displayedRowRef.current.children);

        if (!children.length) return;
        const calculatedItemsWidth: Record<string, number> = {};

        children.forEach((child, index) => {
            calculatedItemsWidth[props.items[index].id] = Math.floor(child.getBoundingClientRect().width);
        });

        return calculatedItemsWidth;
    };

    useLayoutEffectSafeForSsr(() => {
        if (isChanged || !itemsWidth) {
            const newItemsWidth = getItemsWidth();
            setItemsWidth(newItemsWidth);
            setIsChanged(false);
        }
    });

    useEffect(() => {
        const resizeObserver = new ResizeObserver((entries) =>
            requestAnimationFrame(() => {
                if (!Array.isArray(entries) || !entries.length) return;
                setIsChanged(true);
            }));

        resizeObserver.observe(displayedRowRef.current);
        resizeObserver.observe(wrapperRef.current);
        return () => {
            resizeObserver.disconnect();
        };
    }, []);

    const renderItems = () => {
        if (isChanged || !itemsWidth) {
            return props.items.map((i) => i.render(i, [], props.items));
        }
        const wrapperWidth = wrapperRef?.current ? Math.floor(wrapperRef.current.getBoundingClientRect().width) : 0;

        const measuredItems = measureAdaptiveItems(props.items, wrapperWidth, itemsWidth, itemsGap);
        return measuredItems.displayed.map((i) => i.render(i, measuredItems.hidden, measuredItems.displayed));
    };

    let styles;

    if (props.itemsGap) {
        styles = {
            columnGap: `${props.itemsGap}px`,
        };
    }

    return (
        <div { ...props.rawProps } className={ cx(props.cx, css.mainWrapper) } ref={ wrapperRef }>
            <div className={ css.itemsWrapper } style={ styles } ref={ displayedRowRef }>{ renderItems() }</div>
        </div>
    );
}
