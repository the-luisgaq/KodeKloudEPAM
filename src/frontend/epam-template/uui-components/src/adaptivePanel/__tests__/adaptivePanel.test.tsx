import { AdaptivePanel } from '../AdaptivePanel';
import { AdaptiveItemProps } from '../types';
import {
    renderSnapshotWithContextAsync, renderWithContextAsync, mockAdaptivePanelLayout, screen,
} from '@epam/uui-test-utils';
import React from 'react';

type TestItemType = AdaptiveItemProps<{ data?: { label: string } }>;

function getNItems(baseId: string, n: number, priority: number = 1): TestItemType[] {
    const genItem = (params: Pick<TestItemType, 'id' | 'priority'>): TestItemType => {
        return {
            id: params.id,
            priority: params.priority,
            collapsedContainer: false,
            data: { label: `Item ${params.id}` },
            render: (item) => <div data-testid="adaptive-item" data-label={ item.data.label } key={ item.id } />,
        };
    };
    const items: TestItemType[] = [];
    for (let i = 0; i < n; i++) {
        items.push(genItem({ id: `${baseId}-${i}`, priority }));
    }
    return items;
}

async function setupAdaptivePanel({ width, itemWidth, itemsGap }: { width: number; itemWidth: number; itemsGap?: number }) {
    const genContainerItem = ({ id, priority }: Pick<TestItemType, 'id' | 'priority'>): TestItemType => {
        return {
            id,
            priority,
            collapsedContainer: true,
            data: { label: `Collapsed Container Item ${id}` },
            render: (item, hiddenItems) => (
                <div key={ id } data-testid="adaptive-item-cc" data-label={ item.data.label } data-hiddenitems={ hiddenItems.map((i) => i.id).join(',') } />
            ),
        };
    };
    const items: TestItemType[] = [
        ...getNItems('p1', 3, 1),
        ...getNItems('p2', 3, 3),
        ...getNItems('p3', 3, 5),
        genContainerItem({ id: 'wide-screen', priority: 2 }),
        genContainerItem({ id: 'medium-screen', priority: 4 }),
        genContainerItem({ id: 'small-screen', priority: 6 }),
    ];
    mockAdaptivePanelLayout({ width, itemWidth });
    const result = await renderWithContextAsync(<AdaptivePanel items={ items } rawProps={ { 'data-testid': 'adaptive-panel' } } itemsGap={ itemsGap } />);
    const visibleItems = screen
        .queryAllByTestId('adaptive-item')
        .map((i) => i.getAttribute('data-label'))
        .join(',');
    const containerItem = screen.queryByTestId('adaptive-item-cc');
    const hiddenItems = containerItem.getAttribute('data-hiddenitems');
    return {
        result,
        visibleItems,
        hiddenItems,
        containerItem,
    };
}

describe('AdaptivePanel', () => {
    it('should render no items provided', async () => {
        mockAdaptivePanelLayout({ width: 500, itemWidth: 20 });
        const tree = await renderSnapshotWithContextAsync(<AdaptivePanel rawProps={ { 'data-testid': 'adaptive-panel' } } items={ [] } />);
        expect(tree).toMatchSnapshot();
    });

    it('should render some items', async () => {
        mockAdaptivePanelLayout({ width: 500, itemWidth: 20 });
        const tree = await renderSnapshotWithContextAsync(<AdaptivePanel rawProps={ { 'data-testid': 'adaptive-panel' } } items={ getNItems('snapshot', 5) } />);
        expect(tree).toMatchSnapshot();
    });

    it('should render some items with itemsGap', async () => {
        mockAdaptivePanelLayout({ width: 500, itemWidth: 20 });
        const tree = await renderSnapshotWithContextAsync(<AdaptivePanel rawProps={ { 'data-testid': 'adaptive-panel' } } itemsGap={ 6 } items={ getNItems('snapshot', 5) } />);
        expect(tree).toMatchSnapshot();
    });

    it('should properly collapse on WIDE screen', async () => {
        // 7 ( 6 items + 1 collapsed container) * 20 width = 140 Total width;
        const res = await setupAdaptivePanel({ width: 140, itemWidth: 20 });
        expect(res.visibleItems).toEqual('Item p2-0,Item p2-1,Item p2-2,Item p3-0,Item p3-1,Item p3-2');
        expect(res.containerItem).toBeInTheDocument();
        expect(res.containerItem.getAttribute('data-label')).toEqual('Collapsed Container Item wide-screen');
        expect(res.hiddenItems).toEqual('p1-0,p1-1,p1-2');
    });

    it('should properly collapse on WIDE screen with itemsGap', async () => {
        // 7 ( 6 items + 1 collapsed container) * 20 width + 6 gaps * 4px = 140+24 = 164 Total width;
        const res = await setupAdaptivePanel({ width: 164, itemWidth: 20, itemsGap: 4 });
        expect(res.visibleItems).toEqual('Item p2-0,Item p2-1,Item p2-2,Item p3-0,Item p3-1,Item p3-2');
        expect(res.containerItem).toBeInTheDocument();
        expect(res.containerItem.getAttribute('data-label')).toEqual('Collapsed Container Item wide-screen');
        expect(res.hiddenItems).toEqual('p1-0,p1-1,p1-2');
    });

    it('should properly collapse on MEDIUM screen', async () => {
        // 4 ( 3 items + 1 collapsed container) * 20 width = 80 Total width;
        const res = await setupAdaptivePanel({ width: 80, itemWidth: 20 });
        expect(res.visibleItems).toEqual('Item p3-0,Item p3-1,Item p3-2');
        expect(res.containerItem).toBeInTheDocument();
        expect(res.containerItem.getAttribute('data-label')).toEqual('Collapsed Container Item medium-screen');
        expect(res.hiddenItems).toEqual('p1-0,p1-1,p1-2,p2-0,p2-1,p2-2');
    });

    it('should properly collapse on MEDIUM screen with itemsGap', async () => {
        // 4 ( 3 items + 1 collapsed container) * 20 width + (3 gap * 4px )= 80 + 12 = 92 Total width;
        const res = await setupAdaptivePanel({ width: 92, itemWidth: 20, itemsGap: 4 });
        expect(res.visibleItems).toEqual('Item p3-0,Item p3-1,Item p3-2');
        expect(res.containerItem).toBeInTheDocument();
        expect(res.containerItem.getAttribute('data-label')).toEqual('Collapsed Container Item medium-screen');
        expect(res.hiddenItems).toEqual('p1-0,p1-1,p1-2,p2-0,p2-1,p2-2');
    });

    it('should properly collapse on SMALL screen', async () => {
        // 1 collapsed container * 20 width = 20 Total width;
        const res = await setupAdaptivePanel({ width: 20, itemWidth: 20 });
        expect(res.visibleItems).toEqual('');
        expect(res.containerItem).toBeInTheDocument();
        expect(res.containerItem.getAttribute('data-label')).toEqual('Collapsed Container Item small-screen');
        expect(res.hiddenItems).toEqual('p1-0,p1-1,p1-2,p2-0,p2-1,p2-2,p3-0,p3-1,p3-2');
    });

    it('should properly collapse on SMALL screen with itemsGap', async () => {
        // 1 collapsed container * 20 width = 20 Total width; (itemsGap doesn't have effect on the last item or on the collapsed container)
        const res = await setupAdaptivePanel({ width: 20, itemWidth: 20, itemsGap: 4 });
        expect(res.visibleItems).toEqual('');
        expect(res.containerItem).toBeInTheDocument();
        expect(res.containerItem.getAttribute('data-label')).toEqual('Collapsed Container Item small-screen');
        expect(res.hiddenItems).toEqual('p1-0,p1-1,p1-2,p2-0,p2-1,p2-2,p3-0,p3-1,p3-2');
    });
});
