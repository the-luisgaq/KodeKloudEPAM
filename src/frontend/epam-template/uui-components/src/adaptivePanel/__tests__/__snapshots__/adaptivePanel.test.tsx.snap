// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdaptivePanel should render no items provided 1`] = `
<DocumentFragment>
  <div
    class="mainWrapper"
    data-testid="adaptive-panel"
  >
    <div
      class="itemsWrapper"
    />
  </div>
</DocumentFragment>
`;

exports[`AdaptivePanel should render some items 1`] = `
<DocumentFragment>
  <div
    class="mainWrapper"
    data-testid="adaptive-panel"
  >
    <div
      class="itemsWrapper"
    >
      <div
        data-label="Item snapshot-0"
        data-testid="adaptive-item"
      />
      <div
        data-label="Item snapshot-1"
        data-testid="adaptive-item"
      />
      <div
        data-label="Item snapshot-2"
        data-testid="adaptive-item"
      />
      <div
        data-label="Item snapshot-3"
        data-testid="adaptive-item"
      />
      <div
        data-label="Item snapshot-4"
        data-testid="adaptive-item"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`AdaptivePanel should render some items with itemsGap 1`] = `
<DocumentFragment>
  <div
    class="mainWrapper"
    data-testid="adaptive-panel"
  >
    <div
      class="itemsWrapper"
      style="column-gap: 6px;"
    >
      <div
        data-label="Item snapshot-0"
        data-testid="adaptive-item"
      />
      <div
        data-label="Item snapshot-1"
        data-testid="adaptive-item"
      />
      <div
        data-label="Item snapshot-2"
        data-testid="adaptive-item"
      />
      <div
        data-label="Item snapshot-3"
        data-testid="adaptive-item"
      />
      <div
        data-label="Item snapshot-4"
        data-testid="adaptive-item"
      />
    </div>
  </div>
</DocumentFragment>
`;
